ee/modules/021-cni-cilium/hooks/ee/types.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2024 Flant JSC
+Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+*/
+
+package ee
+
+import (
+	"crypto/sha256"
+	"sort"
+)
+
+var hashFunc func(key string) string
+
+func init() {
+	hashFunc = newHashFunc()
+}
+
+type EgressGatewayFailoverGroupState struct {
+	Name               string            `json:"name"`
+	DesiredMasterNode  string            `json:"desiredMasterNode"`
+	AllNodes           []string          `json:"allNodes"`
+	ReadyNodes         []string          `json:"readyNodes"`
+	CurrentMasterNodes []string          `json:"currentMasterNodes"`
+	NodeSelector       map[string]string `json:"nodeSelector"`
+}
+
+func (egfg *EgressGatewayFailoverGroupState) evaluateDesiredMasterNode() string {
+	if len(egfg.ReadyNodes) == 0 {
+		return ""
+	}
+
+	// Algorithm: If the readyNodes slice is empty, then desiredMaster is not calculated. For each of readyNode,
+	// calculate the SHA256 hash of (EgressGatewayGroup.name + Node.Name)
+	// Add the hashes to an array and sort it. Select the zeroth element from the resulting slice.
+
+	// map[<hash>]<node name>
+	nodeHashes := make(map[string]string)
```

Comment:
There is nice trick in metallb
https://github.com/metallb/metallb/blob/e419ead4bd9a3e50dc385ef44f234e31410ddd50/speaker/layer2_controller.go#L117